client/landing/login/store.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * External dependencies
+ */
+import { createStore, applyMiddleware, compose } from 'redux';
+import thunkMiddleware from 'redux-thunk';
+
+/**
+ * Internal dependencies
+ */
+import wpcomApiMiddleware from 'state/data-layer/wpcom-api-middleware';
+import { reducer as httpData, enhancer as httpDataEnhancer } from 'state/data-layer/http-data';
+import { combineReducers } from 'state/utils';
+import application from 'state/application/reducer';
+import currentUser from 'state/current-user/reducer';
+import documentHead from 'state/document-head/reducer';
+import login from 'state/login/reducer';
+import language from 'state/ui/language/reducer';
+import route from 'state/ui/route/reducer';
+import masterbarVisibility from 'state/ui/masterbar-visibility/reducer';
+import section from 'state/ui/section/reducer';
+import layoutFocus from 'state/ui/layout-focus/reducer';
+import notices from 'state/notices/reducer';
+import i18n from 'state/i18n/reducer';
+
+// Create Redux store
+const reducer = combineReducers( {
+	application,
+	currentUser,
```

Comment:
You also need to add the `users` reducer for `currentUser` to work. There is only `state.currentUser.id`, the rest of the user object is in `state.users.items[ state.currentUser.id ]`. I'd like to change that in the future, I even have a WIP branch for that, but it's not trivial.

But the user is not initialized anyway -- the login page always shows the logged-out version. Because the [user initialization code](https://github.com/Automattic/wp-calypso/blob/master/client/boot/app.js#L38-L45) is missing in the login entrypoint.

I think that showing just one version of the login page, no matter whether user is already logged in or not, is actually a good thing to do. It would simplify the login page a lot. The only observable difference between them right now is the masterbar.

But that should probably be a separate and intentional patch.

Reply:
Interesting, that is indeed confusing.

It did remind me of something, though: the user should get redirected if they're already logged in, so there shouldn't be any need for this reducer in the first place inside of the login section. I'm attempting to remove it and running through the E2E test suite to see if that works.

Reply:
> the user should get redirected if they're already logged in

That currently doesn't happen though. And for the case when I want to log in as someone else, a more friendly UI would be to display a "Already logged in as X.Y. Is it you?" warning.

Reply:
The first attempt failed, since the `getCurrentUserId` selector was returning `undefined`, and there are places in the code where `null` is explicitly expected for logged out users. Modified the `getCurrentUserId` selector to default to `null`. Waiting for E2E results to come back.

Reply:
> That currently doesn't happen though. And for the case when I want to log in as someone else, a more friendly UI would be to display a "Already logged in as X.Y. Is it you?" warning.

In the midst of all the changes I got confused. You're right, that's not the current behaviour.

So as I see it, we have some options here:
1 - You get the `log-in` page without any information on whether you're already logged in.
2 - (current) You get the `log-in` page with the logged in masterbar on top, that shows a not so obvious indication you're already logged-in
3 - You get the `log-in` page with the logged in masterbar on top as well as a clearer indication you're already logged in, e.g. "Already logged in as X.Y. Is it you?"

As I understand it, you're arguing for option 2 for now, with option 3 in a later PR?

In the meantime, I'll run some tests to see what the difference in bundle size would be between having user initialization code and not.

Reply:
So, supporting the logged in layout (`client/layout/index.jsx`) would pretty much negate all the savings in this PR, as it pulls in the world and requires a bunch more reducers to be set up.

The alternative would be to create a customized layout which is essentially the logged out layout + the logged in masterbar. That would likely result in an increase of 1-2KB vs. always logged out because of the extra user initialization and user reducers.

How do you want me to move forward, @jsnajdr ?

Reply:
I see two good options:
1. showing exactly the same login page for both logged in and logged out users. The masterbar doesn't indicate logged-in status. A downside is that already logged-in user is in a dead end here: there are no links to a logged-in view from the login page. The logged-in masterbar provided these links.
2. redirecting automatically to a logged-in home page. That's what Dropbox does, for example. The downside is that I can't easily login to another account. I need to go through the `/me` section and click `Sign Out` there. Few more clicks and you need to know where the "Sign Out" is.

Good amendment to option 1 would be detecting that I'm already logged in and offering to proceed or to use another account. That's what Google does. We don't need to bundle the huge `layout/logged-in` to achieve that: just issuing the `/me` request and storing the result in `state.currentUser` should be sufficient. 1-2KB is a good code size estimate.

I don't have enough domain and historical knowledge about our login to make recommendations. Let's ping @stephanethomas and @scruffian to see if they have any insights that could help.

Reply:
I think @westi had opinions on these flows too.

Reply:
> This PR appears to work, but I'm not sure how to test the functionality extensively, beyond attempting login with 2FA and generating/using a magic link.

Here are some additional scenario you may want to consider (I tried to link pull requests with testing instructions that could be useful):

* Using Google to log in (aka. social login)
* Testing login in [other languages](https://github.com/Automattic/wp-calypso/pull/21118) (e.g. [French](https://wordpress.com/log-in/fr))
* Testing login with a redirect url
* Testing login from a [private site](https://github.com/Automattic/wp-calypso/pull/16590)
* Testing login from the desktop app
* Testing login with different [signup flows](https://github.com/Automattic/wp-calypso/pull/22983)

You could also test login flows with other OAuth clients such as [WooCommerce](https://github.com/Automattic/wp-calypso/pull/17699), or Jetpack.

> showing exactly the same login page for both logged in and logged out users.

I'm afraid I don't remember the use case for allowing a logged-in user to log into another account, maybe @scruffian will.

Reply:
Thanks, @stephanethomas! Testing all of those seems like a really good idea.

@scruffian @westi Do you have any opinions on the use case for allowing a logged-in user to log into another account? We're trying to decide between @jsnajdr's two options, which I'm copying below:

> I see two good options:
> 
> 1. showing exactly the same login page for both logged in and logged out users. The masterbar doesn't indicate logged-in status. A downside is that already logged-in user is in a dead end here: there are no links to a logged-in view from the login page. The logged-in masterbar provided these links.
> 2. redirecting automatically to a logged-in home page. That's what Dropbox does, for example. The downside is that I can't easily login to another account. I need to go through the `/me` section and click `Sign Out` there. Few more clicks and you need to know where the "Sign Out" is.
> 
> Good amendment to option 1 would be detecting that I'm already logged in and offering to proceed or to use another account. That's what Google does. We don't need to bundle the huge `layout/logged-in` to achieve that: just issuing the `/me` request and storing the result in `state.currentUser` should be sufficient. 1-2KB is a good code size estimate.